src/java.base/share/classes/jdk/internal/misc/MainMethodFinder.java

```diff
@@ -0,0 +1,171 @@
+                // no arguments or one string array argument.
+                if ("main".equals(method.getName()) &&
+                        !isPrivate(method) &&
+                        correctArgs(method) &&
+                        // Only statics in the declaring class
+                        (!isStatic(method) || declc == refc)
+                ) {
+                    mains.add(method);
+                }
+            }
+
+            gatherMains(declc, refc.getSuperclass(), mains);
+        }
+    }
+
+    /**
+     * Comparator for two methods.
+     * Priority order is;
+     * static < non-static,
+     * public < non-public,
+     * string arg < no arg and
+     * sub-class < super-class.
+     *
+     * @param a  first method
+     * @param b  second method
+     *
+     * @return -1, 0 or 1 to represent higher priority. equals priority or lesser priority.
+     */
+    private static int compareMethods(Method a, Method b) {
+        int aMods = a.getModifiers();
+        int bMods = b.getModifiers();
+        boolean aIsStatic = Modifier.isStatic(aMods);
+        boolean bIsStatic = Modifier.isStatic(bMods);
+
+        if (aIsStatic && !bIsStatic) {
+            return -1;
+        } else if (bIsStatic && !aIsStatic) {
+            return 1;
+        }
+
+        boolean aIsPublic = Modifier.isPublic(aMods);
+        boolean bIsPublic = Modifier.isPublic(bMods);
+
+        if (aIsPublic && !bIsPublic) {
+            return -1;
+        } else if (bIsPublic && !aIsPublic) {
+            return 1;
+        }
+
+        int aCount = a.getParameterCount();
+        int bCount = b.getParameterCount();
+
+        if (aCount > bCount) {
+            return -1;
+        } else if (bCount > aCount) {
+            return 1;
+        }
+
+        Class<?> aClass = a.getDeclaringClass();
+        Class<?> bClass = b.getDeclaringClass();
+
+        if (bClass.isAssignableFrom(aClass)) {
+            return -1;
+        } else if (bClass.isAssignableFrom(aClass)) {
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /**
+     * {@return priority main method or null if none found}
+     *
+     * @param mainClass main class
+     *
+     * @throws NoSuchMethodException when not preview and no method found
+     */
+    public static Method findMainMethod(Class<?> mainClass) throws NoSuchMethodException {
+        try {
+            Method mainMethod = mainClass.getMethod("main", String[].class);
+            int mods = mainMethod.getModifiers();
+
+            if (Modifier.isStatic(mods) && mainMethod.getDeclaringClass() != mainClass) {
+                System.err.println("WARNING: static main in super class will be deprecated.");
+            }
+
+            return mainMethod;
+        } catch (NoSuchMethodException nsme) {
+            if (!PreviewFeatures.isEnabled()) {
+                throw nsme;
+            }
+
+            List<Method> mains = new ArrayList<>();
+            gatherMains(mainClass, mainClass, mains);
```

Comment:
Instead of gatherMains, did you consider first looking for static main(String[], then static main()? Asking because I expected to only see the walk up the hierarchy when looking for an instance main.

Reply:
99.99% of the time it will be a single method in a shallow hierarchy, so cost its low. Only reason I broke out public static main was to ensure performance for existing code was the same.